woocommerce/woocommerce-gateway-stripe
# Remount UPE when removed by updated checkout flow

# Changes proposed in this Pull Request:

Fixes #1731 

This PR adds the functionality to remount the UPE when it gets removed during an update to the checkout. This issue was discovered while testing the new UPE flow with multiple shipping options and other non-UPE payment methods enabled (Sofort or Sepa for example). When updating the shipping method, the payments area was getting cleared and this was removing the previously mounted UPE. The exact reason for this was that the other [Stripe payment methods output the amount](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/5e7a80b1a026b3d56f82cc4ada5b487809c507f3/includes/payment-methods/class-wc-gateway-stripe-sepa.php#L263) as part of their fragment markup. Since this amount was different after changing the shipping option, [the payments area was cleared](https://github.com/woocommerce/woocommerce/blob/2445f7b3033afb78f1d354f0f6467b4266a85504/assets/js/frontend/checkout.js#L375) leaving UPE unmounted. 

This update modifies the `updated_checkout` event handler and `mountUPEElement` functions to remount the UPE in the case that it was removed. As part of that update, some logic was adjusted now to only create the UPE once, and to unmount and remount the element as needed. Since this remounting process was causing more loading scenarios, some of the logic around displaying the loading icons was adjusted as well. The two changes in this area was to apply the `blockUI` call to the parent of the element container and to call `unblockUI` once the element is ready (to prevent a period of empty space while the iframe loads).

One thing I toyed with was removing the `data-amount` attributes from the other payment methods. I could not find anywhere that they were used and upon removing them everything still worked as expected. I opted to leave them in place for two reasons:
- There is a chance they are being used by outside tools.
- Even if we ensure our plugin does not cause a reload of fragments, we can't be sure other plugins won't. So this update would be necessary either way.

If there is a consensus that we should remove these amount values this can be easily added in to this PR.

# Testing instructions

Steps to reproduce the behavior:
- Start with connected Stripe store and running this branch.
- Set `_wcstripe_feature_upe` to 1 in the wp_options table.
- In the Stripe Payments settings, enable the new checkout experience.
- Make sure that at least 1 legacy payment method is active (ex: Sofort) in addition to UPE Credit card
- Add at least 2 shipping options (ex: Flat Rate - $10 and Free Shipping)
- As a customer add a product to the cart
- Go to the checkout and confirm the UPE and legacy payment method loaded and displayed.
- Change the shipping option so that the price will adjust.
- See that the stripe UPE form is properly remounted after the shipping option change.
- Complete the checkout successfully for the correct price + shipping.
- As merchant, confirm order was processed for correct total.
- When reviewing order as merchant, check the transaction details on Stripe to confirm the amount charged was correct.

**Before**
![Stripe-UPE-Before](https://user-images.githubusercontent.com/68524302/130851461-c2e61d7b-4b47-42c2-8852-b319e6eeada0.gif)

**After**
![Stripe-UPE-After](https://user-images.githubusercontent.com/68524302/130851488-fa2dbe65-28bc-42ea-afe5-c963aa6ef896.gif)

-------------------
